docs/research/ha/RATIONALE.md

```diff
@@ -0,0 +1,446 @@
+
+- [High Availability with s3gw](#high-availability-with-s3gw)
+  - [Active-Active pipelines](#active-active-pipelines)
+  - [Active-Passive pipelines](#active-passive-pipelines)
+  - [One Active pipeline](#one-active-pipeline)
+  - [Investigation's Rationale](#investigations-rationale)
+  - [Failure cases](#failure-cases)
+    - [radosgw's POD failure](#radosgws-pod-failure)
+    - [Cluster's node failure](#clusters-node-failure)
+    - [radosgw's failure due to a bug not related to a certain input pattern](#radosgws-failure-due-to-a-bug-not-related-to-a-certain-input-pattern)
+    - [radosgw's failure due to a bug related to a certain input pattern](#radosgws-failure-due-to-a-bug-related-to-a-certain-input-pattern)
+    - [PV Data corruption at application level due to radosgw's bug](#pv-data-corruption-at-application-level-due-to-radosgws-bug)
+    - [PV Data corruption at application level due to radosgw's anomalous exit](#pv-data-corruption-at-application-level-due-to-radosgws-anomalous-exit)
+  - [Measuring s3gw failures on Kubernetes](#measuring-s3gw-failures-on-kubernetes)
+  - [The s3gw Probe](#the-s3gw-probe)
+  - [Tested scenarios](#tested-scenarios)
+    - [EXIT-1, 10 measures](#exit-1-10-measures)
+    - [EXIT-0, 10 measures](#exit-0-10-measures)
+    - [Considerations on the tests performed](#considerations-on-the-tests-performed)
+
+We want to investigate what *High Availability* - HA - means for a project like
+the s3gw.
+
+If we identify the meaning of HA as the ability to have N independent pipelines
+so that the failure of some of these does not affect the user's operations,
+this is something that could be not easy to achieve with the s3gw.
+
+With pipeline we mean all the chain from the ingress to the `PV`:
+
+```mermaid
+flowchart LR
+    ingress(ingress) ==> radosgw(radosgw) ==> PV((PV))
+```
+
+The difficulty with a project like the s3gw is due to the fact that basically
+we have one process: the `radosgw`, that has the *exclusive* access to one resource:
+the Kubernetes `PV` where the S3 buckets and objects are persisted.
+
+Speaking in theory, 3 HA models are possible with s3gw:
+
+1. **Active-Active** pipelines
+2. **Active-Passive** pipelines
+3. **One Active** pipeline (no Passive pipelines preloaded)
+
+## Active-Active pipelines
+
+This is the model that guarantees the best HA characteristics.
+The *Active-Active* model must implement *true independent pipelines*
+where all the pieces are replicated.
+
+The immediate consequence of this statement is that every pipeline being part of
+the same logical s3gw service must bind to a different `PV`; one per `radosgw` process.
+All the `PV`s for the same logical s3gw, must therefore, sync their data.
+
+This implies that with this model, the s3gw must implement an efficient replication
+model across all the pipelines being part of the same logical service.
+
+Under these conditions, a client accessing the S3 service would experience a failure
```

Comment:
No, it might not mean that; the ingress might still mask it and automatically retry. (That's another thing to consider: most S3 protocol libraries are built on the assumption of an unreliable network, and so a single non-repeatable failure might not show to the user, but just be (silently or not) retried and on it goes without anyone knowing beyond a blip in high latency, unless it happens too often. This is a property of the S3 protocol that makes it a bit easier for us to achieve, we hope.)